### PR TITLE
Auto Commands and paths

### PR DESCRIPTION
**Pull request by EliyaFishman**
Now it no longer resets all the time the path when tuning and allows for hotswaping trajectories

### CODE REVIEW

**Comment by EliyaFishman** [comment]
Generated paths still need to be tested with on the robot with falcon dashboard so we can know that the max velocity, acceleration, and constraints are good

File: src/main/java/frc/robot/Robot.java [hunk 1]
```diff
@@ -63,9 +68,12 @@ public void robotInit() {
         dashboardDataContainer = new DashboardDataContainer();
 
         autoChooser = new SendableChooser<>();
-        // autoChooser.setDefaultOption(name, object);
-        // autoChooser.addOption(name, object);
-
+        autoChooser.setDefaultOption("Simple Auto", new SimpleAuto());
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/Robot.java:71]: add another SendableChooser for start pose

File: src/main/java/frc/robot/autonomus/RendezvousAuto.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import frc.robot.motion_profiling.FollowPath;
+import frc.robot.subsystems.drivetrain.RotateDrivetrain;
+
+import static frc.robot.motion_profiling.AutoPath.FacingPowerPortToMiddleField;
+import static frc.robot.motion_profiling.AutoPath.RightOfPortToMiddleField;
+
+/**
+ * This auto command shoots 3 cells, and then goes to collect cells from the RENDEZVOUS POINT.
+ */
+public class RendezvousAuto extends SequentialCommandGroup {
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/autonomus/RendezvousAuto.java]: RendezvousAuto what is that name?

**Comment by EliyaFishman** [comment]
[Create Auto Command groups](https://trello.com/c/3T7xL1aP/131-create-auto-command-groups)

File: src/main/java/frc/robot/constants/robots/RobotA.java [hunk 3]
```diff
@@ -117,6 +117,10 @@ public RobotA() {
         controlConstants.visionRotationSettings = new PIDSettings(0, 0, 0, 0, 0);
         controlConstants.drivetrainRotateSettings = new PIDSettings(0, 0, 0, 0, 0);
 
+        // Auto Constants
+        autoConstants.kSimpleAutoPower = 0.5;
+        autoConstants.kSimpleAutoTimeout = 0.2;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/constants/robots/RobotA.java]: autoConstants.kSimpleAutoTimeout = 0.2;
rise to 0.85

File: src/main/java/frc/robot/constants/robots/RobotA.java [hunk 4]
```diff
@@ -117,6 +117,10 @@ public RobotA() {
         controlConstants.visionRotationSettings = new PIDSettings(0, 0, 0, 0, 0);
         controlConstants.drivetrainRotateSettings = new PIDSettings(0, 0, 0, 0, 0);
 
+        // Auto Constants
+        autoConstants.kSimpleAutoPower = 0.5;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/constants/robots/RobotA.java]: autoConstants.kSimpleAutoPower = 0.5;
lower to 0.3

File: src/main/java/frc/robot/motion_profiling/FollowPath.java [hunk 5]
```diff
@@ -131,6 +139,17 @@ private FollowPath(Trajectory trajectory,
...
     public void initialize() {
+        if (isTuning) {
+            drivetrain.resetOdometry(m_trajectory.getInitialPose());
+            try {
+                String serializedTrajectory =
+                    SmartDashboard.getString(
+                        "Falcon/Serialized Trajectory", TrajectoryUtil.serializeTrajectory(m_trajectory));
+                m_trajectory = TrajectoryUtil.deserializeTrajectory(serializedTrajectory);
+            } catch (JsonProcessingException e) {
+                DriverStationLogger.logErrorToDS("Failed to deserialize trajectory");
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/motion_profiling/FollowPath.java:150]: can you add path name?

File: src/main/java/frc/robot/motion_profiling/FollowPath.java [hunk 6]
```diff
@@ -150,11 +171,8 @@ public void execute() {
         double curTime = m_timer.get();
         double dt = curTime - m_prevTime;
         var curState = m_trajectory.sample(curTime);
-        if (isTuning) {
-            drivetrain.resetOdometry(m_trajectory.getInitialPose());
-            SmartDashboard.putNumber("Falcon/PathX", curState.poseMeters.getTranslation().getX());
-            SmartDashboard.putNumber("Falcon/PathY", curState.poseMeters.getTranslation().getY());
-        }
+        SmartDashboard.putNumber("Falcon/PathX", curState.poseMeters.getTranslation().getX());
+        SmartDashboard.putNumber("Falcon/PathY", curState.poseMeters.getTranslation().getY());
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/motion_profiling/FollowPath.java]: SmartDashboard.putNumber("Falcon/PathX", curState.poseMeters.getTranslation().getX());
        SmartDashboard.putNumber("Falcon/PathY", curState.poseMeters.getTranslation().getY());
make one method: void updateXY()

File: src/main/java/frc/robot/autonomus/SimpleAuto.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    public SimpleAuto() {
+        this(robotConstants.autoConstants.kSimpleAutoTimeout);
+    }
+
+    @Override
+    public void initialize() {
+        timer.reset();
+        timer.start();
+        super.initialize();
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/autonomus/SimpleAuto.java]:
```suggestion
        super.initialize();
        timer.reset();
        timer.start();
```

File: src/main/java/frc/robot/autonomus/SimpleAuto.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+    private final DoubleSupplier timeout;
+    private RunCommand runCommand;
+    private Timer timer;
+
+    public SimpleAuto(DoubleSupplier timeout) {
+        this.timeout = timeout;
+        timer = new Timer();
+        addCommands(new RunCommand(() -> drivetrain.arcadeDrive(0, robotConstants.autoConstants.kSimpleAutoPower),
+            drivetrain).withInterrupt(() -> timer.hasPeriodPassed(timeout.getAsDouble())),
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/autonomus/SimpleAuto.java:21]: why use .withInterrupt? you can use .withTimeout​(double seconds)

File: src/main/java/frc/robot/autonomus/SimpleAuto.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+            new AutoShoot()
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/autonomus/SimpleAuto.java]:
```suggestion
            new AutoShoot(true)
```
we are in auto

File: src/main/java/frc/robot/autonomus/TrenchAuto.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * This auto command shoots 3 balls and then goes to collect more cells from the trench run
+ */
+public class TrenchAuto extends SequentialCommandGroup {
+    public TrenchAuto(StartingPose startingPose) {
+        AutoPath autoPath = startingPose == StartingPose.kFacingPowerPort ?
+            AutoPath.FacingPowerPortToTrenchStart : AutoPath.InLineWithTrenchToTrenchStart;
+        addCommands(
+            new AutoShoot(true),
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/autonomus/TrenchAuto.java:25]: add wait command... of 0.1 seconds. The shooter detects that the ball has been shot but it takes a little bit of time for the actual ball to fly away from the robot. Use: `new WaitCommand​(double timeout)` 
**OR BETTER:** add to AutoShoot that if it is in auto mode (gets true in the constructor) the AutoShoot need to add `new WaitCommand​(double timeout)` at the end of it...

File: src/main/java/frc/robot/autonomus/TrenchAuto.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+            // We get the desired angle by the initial pose rotation in the path.
+            new RotateDrivetrain(() ->
+                autoPath.getPath().getTrajectory().getInitialPose().getRotation().getDegrees()),
+            new FollowPath(autoPath),
+            deadline(
+                new FollowPath(AutoPath.InTrench),
+                new CollectCell()
+            ),
+            new AutoShoot(true)
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/autonomus/TrenchAuto.java]: You have to rotate to the power port before shooting, and you cant shoot under the trench, so rotate, get close to the power port near the start of the trench run and then **shoot!**
**ALSO** the last AutoShoot(true) should be: AutoShoot() without shoot limit of 3 cells (you can make AutoShoot gets the number of cells to shoot).
good luck!

File: src/main/java/frc/robot/autonomus/RendezvousAuto.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+            StartingPose.kFacingRightOfPowerPort ? RightOfPortToMiddleField : FacingPowerPortToMiddleField;
+        addCommands(
+            new AutoShoot(true),
+            new RotateDrivetrain(() ->
+                autoPath.getPath().getTrajectory().getInitialPose().getRotation().getDegrees()),
+            deadline(
+                new FollowPath(autoPath),
+                new CollectCell()
+            ),
+            new AutoShoot(true)
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/autonomus/RendezvousAuto.java]: **see the comments in TrenchAuto first!** 

you cant shoot right away.

**Review by HilelAmiel — CHANGES_REQUESTED** [review]
many

File: src/main/java/frc/robot/autonomus/SimpleAuto.java (hunk 8 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/autonomus/SimpleAuto.java:21]: timeout gets a double. We can't play with the timeout supplier this way as withTimeout is called only one time

**Review by HilelAmiel — APPROVED** [review]

**Pull request merged**